src/VirtualClient/VirtualClient.Dependencies/MsMPIInstallation.cs

```diff
@@ -0,0 +1,94 @@
+        /// Initializes a new instance of the <see cref="MSMPIInstallation"/> class.
+        /// </summary>
+        /// <param name="dependencies">An enumeration of dependencies that can be used for dependency injection.</param>
+        /// <param name="parameters">A series of key value pairs that dictate runtime execution.</param>
+        public MSMPIInstallation(IServiceCollection dependencies, IDictionary<string, IConvertible> parameters)
+            : base(dependencies, parameters)
+        {
+            this.systemManagement = dependencies.GetService<ISystemManagement>();
+            this.packageManager = dependencies.GetService<IPackageManager>();
+            this.processManager = this.systemManagement.ProcessManager;
+        }
+
+        /// <summary>
+        /// Installs the MS-MPI.
+        /// </summary>
+        protected async override Task ExecuteAsync(EventContext telemetryContext, CancellationToken cancellationToken)
+        {
+            if (this.Platform == PlatformID.Win32NT)
+            {
+                DependencyPath workloadPackage = await this.packageManager.GetPlatformSpecificPackageAsync(
+                    this.PackageName, this.Platform, this.CpuArchitecture, CancellationToken.None)
+                    .ConfigureAwait(false);
+
+                string packageDirectory = workloadPackage.Path;
+
+                string osuBenchmarkPath = this.PlatformSpecifics.Combine(packageDirectory, this.PackageName);
+                string lockFile = this.PlatformSpecifics.Combine(packageDirectory, MSMPIInstallation.LockFileName);
+                IFile fileInterface = this.systemManagement.FileSystem.File;
+
+                if (!fileInterface.Exists(lockFile))
+                {
+                    string msmpiBin = this.PlatformSpecifics.Combine(packageDirectory, "bin");
+                    this.SetEnvironmentVariable(EnvironmentVariable.PATH, msmpiBin, EnvironmentVariableTarget.Machine, true);
+
+                    FirewallEntry firewallEntry = new FirewallEntry(
+                    $"Virtual Client: Allow {this.PlatformSpecifics.Combine(msmpiBin, "smpd.exe")}",
+                    $"Allows smpd to Communicate in Multiple Machine Scenario.",
+                    this.PlatformSpecifics.Combine(msmpiBin, "smpd.exe"));
+
+                    await this.systemManagement.FirewallManager.EnableInboundAppAsync(firewallEntry, cancellationToken)
+                        .ConfigureAwait(false);
+
+                    await this.InstallComponentAsync(
+                        this.PlatformSpecifics.Combine(packageDirectory, "msmpisdk.msi"),
+                        telemetryContext,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+
+                    fileInterface.Create(lockFile);
+                }
+            }
+        }
+
+        private async Task InstallComponentAsync(string msiPath, EventContext telemetryContext, CancellationToken cancellationToken)
```

Comment:
We have an extension method for running, logging, throwing altogether in VC component now.